src/owr/OptimisticWithdrawalRecipientPectra.sol

```diff
@@ -0,0 +1,451 @@
+
+    // determine which recipeint is getting paid based on funds to be
+    // distributed
+    uint256 _principalPayout = 0;
+    uint256 _rewardPayout = 0;
+
+    unchecked {
+      // _claimedPrincipalFunds should always be <= amountOfPrincipalStake
+      uint256 principalStakeRemaining = amountOfPrincipalStake - _claimedPrincipalFunds;
+
+      if (_fundsToBeDistributed >= BALANCE_CLASSIFICATION_THRESHOLD && principalStakeRemaining > 0) {
+        if (_fundsToBeDistributed > principalStakeRemaining) {
+          // this means there is reward part of the funds to be
+          // distributed
+          _principalPayout = principalStakeRemaining;
+          // shouldn't underflow
+          _rewardPayout = _fundsToBeDistributed - principalStakeRemaining;
+        } else {
+          // this means there is no reward part of the funds to be
+          // distributed
+          _principalPayout = _fundsToBeDistributed;
+        }
+      } else {
+        _rewardPayout = _fundsToBeDistributed;
+      }
+    }
+
+    {
+      if (_fundsToBeDistributed > type(uint128).max) revert InvalidDistribution_TooLarge();
+      // Write to storage
+      // the principal value
+      // it cannot overflow because _principalPayout < _fundsToBeDistributed
+      if (_principalPayout > 0) claimedPrincipalFunds += uint128(_principalPayout);
+    }
+
+    /// interactions
+
+    // pay outs
+    // earlier tranche recipients may try to re-enter but will cause fn to
+    // revert
+    // when later external calls fail (bc balance is emptied early)
+
+    // pay out principal
+    _payout(principalRecipient, _principalPayout, pullFlowFlag);
+    // pay out reward
+    _payout(rewardRecipient, _rewardPayout, pullFlowFlag);
+
+    if (pullFlowFlag == PULL) {
+      if (_principalPayout > 0 || _rewardPayout > 0) {
+        // Write to storage
+        fundsPendingWithdrawal = uint128(_memoryFundsPendingWithdrawal + _principalPayout + _rewardPayout);
+      }
+    }
+
+    emit DistributeFunds(_principalPayout, _rewardPayout, pullFlowFlag);
+  }
+
+  function _payout(address recipient, uint256 payoutAmount, uint256 pullFlowFlag) internal {
+    if (payoutAmount > 0) {
+      if (pullFlowFlag == PULL) {
+        // Write to Storage
+        pullBalances[recipient] += payoutAmount;
+      } else {
+        recipient.safeTransferETH(payoutAmount);
+      }
+    }
+  }
+
+  function _requestWithdrawal(bytes memory pubkey, uint8 amount, bool _rewards) private {
+    if (pubkey.length != 48) revert InvalidWithdrawal_Failed();
+
+    if (
+      (!_rewards && amount < BALANCE_CLASSIFICATION_THRESHOLD)
+        || (_rewards && amount >= BALANCE_CLASSIFICATION_THRESHOLD)
+    ) {
+      if (_rewards) revert InvalidPectraWithdrawal_Rewards();
+      else revert InvalidPectraWithdrawal_Principal();
+    }
+
+    // Input data has the following layout:
+    //
+    //  +--------+--------+
+    //  | pubkey | amount |
+    //  +--------+--------+
+    //      48       8
```

Comment:
Maybe a stupid question but, are these measured in bits or bytes? 

Because uint8, is 8 bits, representing a max number of 255. but is `bytes memory pubkey` also measured in bits? or is it measured in bytes? A validator pubkey is 96 hex chars, or 48 byte chars, or 384 bits.  

Should `amount` be a uint64? which is 8*8bytes. That seems to make more sense for an amount to withdraw than a number from 0-> 255

Reply:
Yes, it's uint64. An example could be seen in the mock contract where the request is decoded.